Improve game mode support

obrakmann opened the pull request:
This PR adds a game mode manager which allows selecting one of several game modes that can be played on a single map.

Key changes:
* Adds a game mode manager
* Adds a `GameModesInit` actor init, used to restrict an actor's appearance to specific game modes.
* Allows having more than one `LuaScript`

Prerequisite:
* [Time limit lobby option](https://github.com/OpenRA/OpenRA/compare/bleed...obrakmann:timelimit)

Pending follow-ups:
* [UI and editor changes](https://github.com/OpenRA/OpenRA/compare/obrakmann:new-game-modes...obrakmann:new-game-modes-2)
* [Domination game mode](https://github.com/OpenRA/OpenRA/compare/obrakmann:new-game-modes-2...obrakmann:new-game-modes-3)

![screenshot from 2018-11-01 23-51-37](https://user-images.githubusercontent.com/4331210/47884671-f6ca0a80-de31-11e8-9d19-081807805471.png)

obrakmann committed:
Add support for multiple game modes per map

obrakmann committed:
test case

obrakmann committed:
Allow multiple instances of LuaScript

pchote commented:
Maybe i've swung too far towards "Everything is a condition", but this does seem like a good usecase for them.

We could replace the `IVictoryConditionInfo` interface with a `GameMode` trait that lists the metadata (name + condition to grant) for each mode.  The victory condition traits would then implement `ConditionalTrait` and be activated by the conditions granted to the World and Player actors (but may need to go via an ExternalCondition to satisfy the linter... bleh).

This would give us:
* The ability to customize AI behaviour for each game mode (#15750 and followups)
* The ability to maps to define their own game modes that mix and match existing logic (e.g. `ConquestVictoryConditions` + additional logic implemented in a Lua script)
* The ability to turn toggle arbitrary other player/world traits: e.g. grant custom prereqs, override crate behaviour, change factions in the lobby if #3012 is ever finished

What do you think?

obrakmann commented:
Sounds reasonable. Using upgrades/conditions was actually your original suggestion in https://github.com/OpenRA/OpenRA/issues/10367#issuecomment-168314673 as well, but I left it like it was when I saw that neither `World` nor `Player` seem to have a `ConditionsManager` yet. I'll just have to find a nice way to pass those conditions around, so that `SpawnMapActors` can see them.

obrakmann committed:
Add RejectsOrders trait to World to provide a dummy Condition consumer

This can be removed after real users have been added

obrakmann commented:
Updated & rebased. This will now need upgrade rules, which are still missing.

obrakmann commented:
Upgrade rules have been added, this should now be ready.

obrakmann commented:
Rebased. Had to rework the time limit thing because it turned out it didn't actually work. Whoops.  Added some more commits, which should now fix #10367.

obrakmann committed:
Add upgrade rules for GameModeManager addition

obrakmann commented:
As discussed in IRC, reduced this PR's scope to the bare minimum required to ease review.

pchote commented:
IMO we should try and get this merged ASAP after we branch the next playtest to maximize the amount of time to handle the followups (adding game modes, dealing with the custom balance rule overlays).

pchote commented on OpenRA.Mods.Common/Scripting/LuaScript.cs:
```diff
@@ -19,34 +19,37 @@
 		void IWorldLoaded.WorldLoaded(World world, WorldRenderer worldRenderer)
 		{
-			var scripts = info.Scripts ?? Enumerable.Empty<string>();
+			var scripts = IsTraitDisabled ? Enumerable.Empty<string>() : info.Scripts ?? Enumerable.Empty<string>();
```
What happens / what do we want to happen if the trait is enabled after the game starts?

IMO the best solution here is to override `TraitEnabled` and throw an exception if the trait is enabled after the game has started.  The victory condition traits should do the same thing, otherwise e.g. `StrategicProgressWidget` will need to account for the victory trait changing.

pchote commented on OpenRA.Mods.Common/Traits/Player/GameMode.cs:
```diff
@@ -0,0 +1,57 @@
+	// GameModeInit is used to restrict Actors to certain game modes, eg. only showing KotH flags on the koth game mode.
+	// An empty value is interpreted to mean that the actor should appear in all game modes.
+	// Matches against the InternalName field of a GameMode.
+	public class GameModesInit : IActorInit<string[]>
```
I suggested a very similar (but more general) Init in https://github.com/IceReaper/KKnD/issues/22.  Would be good to at least consider this here, but otherwise / at least I think we should expose this init in the new editor actor properties panel.

pchote commented on OpenRA.Mods.Common/Traits/Player/MissionObjectives.cs:
```diff
@@ -168,7 +168,7 @@ void CheckIfGameIsOver(Player player)
 		}
 
-		public void OnPlayerWon(Player player)
+		void INotifyObjectivesUpdated.OnPlayerWon(Player player)
```
Why can this trait stay not-conditional?

obrakmann replied:
I moved the two traitinfo settings to the GameMode trait, so this can now *really* stay unconditional.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoLogic.cs:
```diff
@@ -29,7 +29,7 @@ public GameInfoLogic(Widget widget, World world, IngameInfoPanel activePanel)
 
 			// Objectives/Stats tab
-			var scriptContext = world.WorldActor.TraitOrDefault<LuaScript>();
+			var scriptContext = world.WorldActor.TraitsImplementing<LuaScript>().FirstEnabledTraitOrDefault();
```
This would be better written as
```csharp
var hasError = world.WorldActor.TraitsImplementing<LuaScript>()
	.Any(sc => !sc.IsTraitDisabled && sc.FatalErrorOccurred);
```

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -36,7 +36,7 @@ public IngameMenuLogic(Widget widget, ModData modData, World world, Action onExi
 			menu.Get("MENU_BUTTONS").IsVisible = () => !hideMenu;
 
-			var scriptContext = world.WorldActor.TraitOrDefault<LuaScript>();
+			var scriptContext = world.WorldActor.TraitsImplementing<LuaScript>().FirstEnabledTraitOrDefault();
```
And similarly here too.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/LoadIngamePlayerOrObserverUILogic.cs:
```diff
@@ -62,7 +63,7 @@ public LoadIngamePlayerOrObserverUILogic(Widget widget, World world)
 				if (world.LocalPlayer != null)
 				{
-					var scriptContext = world.WorldActor.TraitOrDefault<LuaScript>();
+					var scriptContext = world.WorldActor.TraitsImplementing<LuaScript>().FirstEnabledTraitOrDefault();
```
And here

pchote commented on OpenRA.Mods.Common/Widgets/StrategicProgressWidget.cs:
```diff
@@ -20,13 +20,15 @@ namespace OpenRA.Mods.Common.Widgets
 		{
 			IsVisible = () => true;
 			this.world = world;
+			svc = world.Players.Select(p => p.PlayerActor.TraitOrDefault<StrategicVictoryConditions>()).FirstOrDefault();
```
This needs to filter out disabled traits

pchote reviewed (commented):
Code looks generally good, just a few observations to start with.
Haven't tested ingame yet.

pchote commented:
The "game mode" label under the map name in the lobby needs to be switched over to the new plumbing - the assumption that the first category is the game mode is no longer valid.

<img width="547" alt="screenshot 2018-12-20 at 12 42 32" src="https://user-images.githubusercontent.com/167819/50254943-e1965300-0454-11e9-8e2c-f490a3c51102.png">

We probably also want to consider renaming the "Conquest" category to something else, and adding a  new game mode filter to the map chooser.

obrakmann committed:
Add an ActorInit to restrict actors to certain game modes

obrakmann committed:
Add a time limit lobby option

obrakmann committed:
Let sidebar timer count backwards if a time limit is set

obrakmann committed:
Add scripting support for TimeLimitManager

obrakmann committed:
Convert Allies-02 to make use of TimeLimit API

obrakmann committed:
Convert "Top o' the World" mission to use TimeLimit API

obrakmann committed:
Disable TimeLimit dropdown on MP missions and Fort Lonestar

obrakmann commented:
I think I'm going to close this to clean up the PR queue. The follow-up branches have all been updated, so if someone wants to pick this up, feel free.

pchote commented:
I definitely still want to see this get in, but probably best to push it to Next + 1 or later now.

obrakmann commented:
Is there still interest in finishing this?
